edgedb/protocol/protocol.pyx

```diff
@@ -219,30 +331,38 @@ cdef class SansIOProtocol:
         allow_capabilities: enums.Capability = enums.Capability.ALL,
         in_dc: BaseCodec,
         out_dc: BaseCodec,
-        allow_re_exec: bint = True,
     ):
         cdef:
             WriteBuffer packet
             WriteBuffer buf
+            WriteBuffer params
             char mtype
-            bint re_exec
+            bint need_parse = False
             object result
             bytes new_cardinality = None
 
-        compilation_flags = enums.CompilationFlag.INJECT_OUTPUT_OBJECT_IDS
-        if inline_typenames:
-            compilation_flags |= enums.CompilationFlag.INJECT_OUTPUT_TYPE_NAMES
-        if inline_typeids:
-            compilation_flags |= enums.CompilationFlag.INJECT_OUTPUT_TYPE_IDS
+        params = self.encode_parse_params(
+            query=query,
+            output_format=output_format,
+            expect_one=expect_one,
+            required_one=required_one,
+            implicit_limit=implicit_limit,
+            inline_typenames=inline_typenames,
+            inline_typeids=inline_typeids,
+            allow_capabilities=allow_capabilities,
+        )
+
+        if isinstance(in_dc, InvalidCodec):
+            if not args and not kwargs:
+                in_dc = NULL_CODEC
```

Comment:
This `NULL_CODEC` guess could happen earlier in `def execute()` before calling `self._parse()`?